### PR TITLE
Pull mobile device metrics

### PR DESCRIPTION
**Pull request by TomSendrovich**

### CODE REVIEW

File: Ginger/Ginger/Drivers/DriversWindows/MobileDriverWindow.xaml.cs [hunk 1]
```diff
@@ -190,6 +317,309 @@ private async void MobileDriverWindow_DriverMessageEvent(object sender, DriverMe
...
+                    }
+                }
+
+                if (mDeviceGeneralInfo.TryGetValue("model", out value))
+                {
+                    if (string.IsNullOrEmpty((string)value))
+                    {
+                        mDeviceDetails.Add(new DeviceInfo("Model:", "N/A", DeviceInfo.eCategory.DeviceDetails));
+
+                        //xModelValueLbl.Content = "N/A";
```
Comment (MeniKadosh1) [inline Ginger/Ginger/Drivers/DriversWindows/MobileDriverWindow.xaml.cs:464]: remove commented out code

File: Ginger/Ginger/Drivers/DriversWindows/MobileDriverWindow.xaml.cs [hunk 2]
```diff
@@ -190,6 +317,309 @@ private async void MobileDriverWindow_DriverMessageEvent(object sender, DriverMe
...
+                    }
+                }
+
+                if (mDeviceBatteryInfo.TryGetValue("power", out battery))
+                {
+                    mDeviceDetails.Add(new DeviceInfo("Device Battery:", battery + "%", DeviceInfo.eCategory.DeviceMetrics));
+                }
+                else
+                {
+                    mDeviceDetails.Add(new DeviceInfo("Battery:", "N/A", DeviceInfo.eCategory.DeviceMetrics));
```
Comment (MeniKadosh1) [inline Ginger/Ginger/Drivers/DriversWindows/MobileDriverWindow.xaml.cs:586]: why "Device Battery" and for N/A it is only "Battery"?

File: Ginger/Ginger/Drivers/DriversWindows/MobileDriverWindow.xaml.cs [hunk 3]
```diff
@@ -190,6 +317,309 @@ private async void MobileDriverWindow_DriverMessageEvent(object sender, DriverMe
...
+
+                if (mDeviceGeneralInfo.TryGetValue("carrierName", out value))
+                {
+                    if (string.IsNullOrEmpty((string)value))
+                    {
+                        mDeviceDetails.Add(new DeviceInfo("Carrier name:", "N/A", DeviceInfo.eCategory.DeviceDetails));
+                    }
+                    else
+                    {
+                        mDeviceDetails.Add(new DeviceInfo("Carrier name:", (string)value, DeviceInfo.eCategory.DeviceDetails));
```
Comment (MeniKadosh1) [inline Ginger/Ginger/Drivers/DriversWindows/MobileDriverWindow.xaml.cs:564]: remove ":" from all details/metrics names

File: Ginger/Ginger/Drivers/DriversWindows/MobileDriverWindow.xaml.cs [hunk 4]
```diff
@@ -190,6 +317,309 @@ private async void MobileDriverWindow_DriverMessageEvent(object sender, DriverMe
...
+                }
+                else
+                {
+                    mDeviceDetails.Add(new DeviceInfo("Battery:", "N/A", DeviceInfo.eCategory.DeviceMetrics));
+                }
+
+            }
+            else
+            {
+                mDeviceDetails.Add(new DeviceInfo("Device ID:", "N/A", DeviceInfo.eCategory.DeviceDetails));
```
Comment (MeniKadosh1) [inline Ginger/Ginger/Drivers/DriversWindows/MobileDriverWindow.xaml.cs:592]: this else is not needed, just causing duplicated list to maintain, you already checking for each field if it exist and if not setting it to "N/A" so why again?

File: Ginger/Ginger/Drivers/DriversWindows/MobileDriverWindow.xaml.cs [hunk 5]
```diff
@@ -190,6 +317,309 @@ private async void MobileDriverWindow_DriverMessageEvent(object sender, DriverMe
...
+
+            if (string.IsNullOrEmpty(mAgent.Name))
+            {
+                //xAgentValueLbl.Content = "N/A";
+                mDeviceDetails.Add(new DeviceInfo("Ginger Agent:", "N/A", DeviceInfo.eCategory.DeviceDetails));
+            }
+            else
+            {
+                //xAgentValueLbl.Content = mAgent.Name;
+                mDeviceDetails.Add(new DeviceInfo("Ginger Agent:", mAgent.Name, DeviceInfo.eCategory.DeviceDetails, "Has extra info"));
```
Comment (MeniKadosh1) [inline Ginger/Ginger/Drivers/DriversWindows/MobileDriverWindow.xaml.cs:616]: "Value" of the Agent need to be the AgentName

File: Ginger/Ginger/Drivers/DriversWindows/MobileDriverWindow.xaml.cs [hunk 6]
```diff
@@ -647,22 +1093,26 @@ public void InitWindowLook()
...
+                    this.Icon = ImageMakerControl.GetImageSource(eImageType.Ios);
                 }
-                this.Title = string.Format("Ginger {0} Device View", mAgent.Name);
+
                 this.Width = 320;
                 this.Height = 650;
                 xMessageLbl.Content = "Connecting to Device...";
 
                 //Configurations
                 SetConfigurationsPanelView(false);
```
Comment (MeniKadosh1) [inline Ginger/Ginger/Drivers/DriversWindows/MobileDriverWindow.xaml.cs:1112]: why it is needed on init? don't we need only the title setup? rest grids view and data can be set only on metrics section expanding

File: Ginger/Ginger/Drivers/DriversWindows/MobileDriverWindow.xaml.cs [hunk 7]
```diff
@@ -1054,9 +1529,96 @@ private void PerformScreenSwipe(eSwipeSide swipeSide, double impact = 1)
...
+                agentEditPage.Width = 800;
+                agentEditPage.Height = 800;
+                ((Grid)((ScrollViewer)agentEditPage.Content).Content).IsEnabled = false;
+                GingerCore.General.LoadGenericWindow(ref genWin, App.MainWindow, eWindowShowStyle.Free, mAgent.Name, agentEditPage);
+            });
+        }
+
+        private void xRamUsageBtn_Click(object sender, RoutedEventArgs e)
+        {
+            OpenPopUpWindow(DictionaryToString(mDeviceMemoryInfo).Replace(", ", Environment.NewLine), "Full device's Memory Information");
```
Comment (MeniKadosh1) [inline Ginger/Ginger/Drivers/DriversWindows/MobileDriverWindow.xaml.cs:1571]: need to change title, this is not "Full device.." but only App Memory and CPU- need to change for both

File: Ginger/Ginger/Drivers/DriversWindows/MobileDriverWindow.xaml.cs [hunk 8]
```diff
@@ -1054,9 +1529,96 @@ private void PerformScreenSwipe(eSwipeSide swipeSide, double impact = 1)
...
+        }
+
+        private void xNetworksBtn_Click(object sender, RoutedEventArgs e)
+        {
+            OpenPopUpWindow(mDeviceNetworkInfo, "Full device's Network Information");
+        }
+
+
+
+        //private void xAgentValueLbl_TargetUpdated(object sender, System.Windows.Data.DataTransferEventArgs e)
```
Comment (MeniKadosh1) [inline Ginger/Ginger/Drivers/DriversWindows/MobileDriverWindow.xaml.cs:1581]: delete commented out code

File: Ginger/Ginger/Drivers/DriversWindows/MobileDriverWindow.xaml.cs [hunk 9]
```diff
@@ -1054,9 +1529,96 @@ private void PerformScreenSwipe(eSwipeSide swipeSide, double impact = 1)
...
+        //    string labelContent = xAgentValueLbl.Content.ToString();
+        //    if (labelContent.Length > 35)
+        //    {
+        //        xAgentValueLbl.Content = labelContent.Substring(0, 35) + "...";
+        //    }
+        //}
+
+        private void xExtraInfoBtn_ExtraInfo_Click(object sender, RoutedEventArgs e)
+        {
+            DeviceInfo deviceInfo = (DeviceInfo)((ucButton)sender).DataContext;
```
Comment (MeniKadosh1) [inline Ginger/Ginger/Drivers/DriversWindows/MobileDriverWindow.xaml.cs:1592]: i didn't get why this method is needed if you already having event handling for each extra info click above.
just add to DeviceInfo field which will store the event and trigger it on click

File: Ginger/Ginger/Drivers/DriversWindows/MobileDriverWindow.xaml.cs [hunk 10]
```diff
@@ -1054,9 +1529,96 @@ private void PerformScreenSwipe(eSwipeSide swipeSide, double impact = 1)
...
+                case "App CPU Usage:":
+                    OpenPopUpWindow(DictionaryToString(mDeviceCPUInfo).Replace(", ", Environment.NewLine), "Full device's CPU Information");
+                    break;
+                case "Ginger Agent:":
+                    Application.Current.Dispatcher.Invoke(() =>
+                    {
+                        AgentEditPage agentEditPage = new AgentEditPage(mAgent);
+                        agentEditPage.Width = 1000;
+                        agentEditPage.Height = 1000;
+                        ((TextBox)agentEditPage.FindName("xAgentNameTextBox")).IsReadOnly = true;
```
Comment (MeniKadosh1) [inline Ginger/Ginger/Drivers/DriversWindows/MobileDriverWindow.xaml.cs:1610]: that is not pretty way for showing in read only- need to have flag on the page to load it as REadOnly and then to handle it inside the page and not externally

File: Ginger/GingerCoreCommon/ReporterLib/UserMsgsPool.cs [hunk 11]
```diff
@@ -310,6 +310,7 @@ public static void LoadUserMsgsPool()
             #endregion Environment Messages
 
             #region Agents/Drivers Messages
+            Reporter.UserMsgsPool.Add(eUserMsgKey.MobileDriverNotConnected, new UserMsg(eUserMsgType.INFO, "Mobile Driver Error", "Mobile driver could not connet to mobile device, please try again.", eUserMsgOption.OK, eUserMsgSelection.None));
```
Comment (MeniKadosh1) [inline Ginger/GingerCoreCommon/ReporterLib/UserMsgsPool.cs:313]: users don't know what is Mobile Driver, set the error to "The connection to the mobile device has failed, please try again."

File: Ginger/GingerCoreNET/ActionsLib/UI/Mobile/ActMobileDevice.cs [hunk 12]
```diff
@@ -199,7 +196,17 @@ public enum eMobileDeviceAction
             [EnumValueDescription("Lock Device")]
             LockDevice,
             [EnumValueDescription("Get Page Source")]
-            GetPageSource,                      
+            GetPageSource,
+            [EnumValueDescription("Get Device's Battery %")]
+            GetDeviceBattery,
+            [EnumValueDescription("Get Device's Network Info")]
+            GetDeviceNetwork,
+            [EnumValueDescription("Get Device's CPU Usage")]
```
Comment (MeniKadosh1) [inline Ginger/GingerCoreNET/ActionsLib/UI/Mobile/ActMobileDevice.cs:204]: need to change to "Get App CPU Usage" also for RAM

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Mobile/Appium/GenericAppiumDriver.cs [hunk 13]
```diff
@@ -961,7 +990,28 @@ private void MobileDeviceActionHandler(ActMobileDevice act)
                             act.Error = "Operation not supported for this mobile OS or application type.";
                         }
                         break;
-
+                    case ActMobileDevice.eMobileDeviceAction.GetDeviceBattery:
+                        AddReturnParamFromDict(GetDeviceBatteryInfo(), act);
```
Comment (MeniKadosh1) [inline Ginger/GingerCoreNET/Drivers/CoreDrivers/Mobile/Appium/GenericAppiumDriver.cs:994]: i'm missing here the implementation of all the methods like "GetDeviceBatteryInfo()" from some reason not seeing it on PR, you will show me on your VS

**Review by MeniKadosh1 — CHANGES_REQUESTED** [review]

File: Ginger/Ginger/Drivers/DriversWindows/DeviceInfo.cs [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+    public class DeviceInfo : INotifyPropertyChanged
+    {
+        public DeviceInfo(string name, string value, eCategory category, string extraInfo = null)
+        {
+            DetailName = name;
+            DetailValue = value;
+            ExtraInfo = extraInfo;
+            Category = category;
+        }
+        public enum eCategory
```
[the remark this answers is not in the document]
Comment (TomSendrovich) [reply Ginger/Ginger/Drivers/DriversWindows/DeviceInfo.cs:20]: Done

File: Ginger/Ginger/Drivers/DriversWindows/MobileDriverWindow.xaml.cs (hunk 9 above)
Comment (TomSendrovich) [reply Ginger/Ginger/Drivers/DriversWindows/MobileDriverWindow.xaml.cs:1592]: I removed to above individual functions and kept this function instead.

File: Ginger/Ginger/Drivers/DriversWindows/MobileDriverWindow.xaml.cs [hunk 15]
```diff
@@ -148,10 +245,40 @@ private async void MobileDriverWindow_DriverMessageEvent(object sender, DriverMe
                             xMessagePnl.Visibility = Visibility.Collapsed;
                             xDeviceScreenshotCanvas.Visibility = Visibility.Visible;
                             xMessageLbl.Content = "Loading Device Screenshot...";
+
+                            //Loading image on details and metrics ucGrid
```
Comment (MeniKadosh1) [inline Ginger/Ginger/Drivers/DriversWindows/MobileDriverWindow.xaml.cs:249]: on device load no need to details with details/metrics controls or data besides getting the needed data for title of the window, rest needs to be done only when used clicks on the "Metrics" icon on tool bar

File: Ginger/Ginger/Drivers/DriversWindows/MobileDriverWindow.xaml.cs [hunk 16]
```diff
@@ -190,6 +317,302 @@ private async void MobileDriverWindow_DriverMessageEvent(object sender, DriverMe
...
+                }
+                else
+                {
+                    mDeviceDetails.Add(new DeviceInfo("Battery:", "N/A", DeviceInfo.eDeviceInfoCategory.Detail));
+                }
+
+            }
+            else
+            {
+                mDeviceDetails.Add(new DeviceInfo("Device ID:", "N/A", DeviceInfo.eDeviceInfoCategory.Detail));
```
Comment (MeniKadosh1) [inline Ginger/Ginger/Drivers/DriversWindows/MobileDriverWindow.xaml.cs:587]: No need to maintain 2 "N/A" set, just add to the IF check both checks if field exist and not null

File: Ginger/GingerCoreNET/Drivers/CoreDrivers/Mobile/Appium/GenericAppiumDriver.cs [hunk 17]
```diff
@@ -144,14 +150,38 @@ public bool IsDeviceConnected
...
         {
             get => LoadDeviceWindow;
         }
 
         private AppiumDriver Driver;//appium 
         private SeleniumDriver mSeleniumDriver;//selenium 
-     
+
+        private Dictionary<string, object> mDeviceGeneralInfo;
+        public Dictionary<string, object> DeviceGeneralInfo { get => mDeviceGeneralInfo; set => mDeviceGeneralInfo = value; }
```
Comment (MeniKadosh1) [inline Ginger/GingerCoreNET/Drivers/CoreDrivers/Mobile/Appium/GenericAppiumDriver.cs:164]: No need the Dic's can return directly from calculating method

**Review by MeniKadosh1 — CHANGES_REQUESTED** [review]

**Review by MeniKadosh1 — APPROVED** [review]

**Pull request merged**